# Detecting Fake News with Python and Machine Learning

To build a model to accurately classify a piece of news as REAL or FAKE.

## About Detecting Fake News with Python
This advanced python project of detecting fake news deals with fake and real news. Using sklearn, we build a TfidfVectorizer on our dataset. Then, we initialize a PassiveAggressive Classifier and fit the model. In the end, the accuracy score and the confusion matrix tell us how well our model fares.

### What is a TfidfVectorizer?
TF (Term Frequency): The number of times a word appears in a document is its Term Frequency. A higher value means a term appears more often than others, and so, the document is a good match when the term is part of the search terms.

IDF (Inverse Document Frequency): Words that occur many times a document, but also occur many times in many others, may be irrelevant. IDF is a measure of how significant a term is in the entire corpus.

The TfidfVectorizer converts a collection of raw documents into a matrix of TF-IDF features.

### What is a PassiveAggressiveClassifier?
Passive Aggressive algorithms are online learning algorithms. Such an algorithm remains passive for a correct classification outcome, and turns aggressive in the event of a miscalculation, updating and adjusting. Unlike most other algorithms, it does not converge. Its purpose is to make updates that correct the loss, causing very little change in the norm of the weight vector.

## The fake news Dataset
The dataset we’ll use for this python project- we’ll call it news.csv. This dataset has a shape of 7796×4. The first column identifies the news, the second and third are the title and text, and the fourth column has labels denoting whether the news is REAL or FAKE.

In [1]:
pip install numpy

Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [3]:
pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.


# Necessary imports

In [4]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

Read the data into the DataFrame, use head() to get the first 5 rows.

In [5]:
df=pd.read_csv("/Users/okguser/Downloads/news.csv")

df.shape
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

And get the labels from the DataFrame.

In [6]:
labels = df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

Split the dataset into training and testing sets.

In [7]:
#x_train, input data for training set
#x_test, contain the features for testing set
#y_train, output data for training set
#y_test, contain the corresponding labels for the testing set
#df['test'], DataFrame column containing the text data or features
#labels, containing the labels corresponding to the text data
#test_size, specifies that how many percentage of the data will be used for testing
#random_state, sets the random seed ensuring reproducibility if rerun the code. The same split will be obtained each time with this seed.
x_train, x_test, y_train, y_test = train_test_split(df['text'], labels, test_size=0.2, random_state = 7)

Initialize a TfidVectorizer with stop words from the English language and a maximum document frequency of 0.7. Stop words are the most common words in a language that are to be filtered out before processing the natural language data. And a TfidfVectorizer turns a collection of raw documents into a matrix of TF-IDF features.

Fit and transform the vectorizer on the train set, and transform the vectorizer on the test set.

In [8]:
#initialize a TfidfVectorizer
#stop_words, removes common english stop words from the text
#max_df, ignores words that appear in more than 70% of the documents
tfidf_vectorizer = TfidfVectorizer(stop_words='english', max_df=0.7)

#fit and transform train set and test set
tfidf_train = tfidf_vectorizer.fit_transform(x_train)
tfidf_test = tfidf_vectorizer.transform(x_test)

Initialize a PassiveAggressiveClassifier. This is. We’ll fit this on tfidf_train and y_train.

Then, we’ll predict on the test set from the TfidfVectorizer and calculate the accuracy with accuracy_score() from sklearn.metrics.

In [9]:
#initialize a PassiveAggressiveClassifier
#maximum of 50 iterations
pac = PassiveAggressiveClassifier(max_iter=50)
#fit the classifier
pac.fit(tfidf_train,y_train)
#predict on the test set and calculate accuracy
y_pred = pac.predict(tfidf_test)
#calculates the accuracy of the predictions by comparing predictied labels with actual labels of the test set
score = accuracy_score(y_test,y_pred)
#print the accuracy score by rounded two decimal places
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.9%


Got an accuracy of 92.9% with this model. Print out a confusion matrix to gain insight into the number of false and true negatives and positives.

In [10]:
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[591,  47],
       [ 43, 586]])

With this model, we have 591 true positives, 586 true negatives, 47 false positives, and 43 false negatives.